### DISCUSSION TITLE
Client-side instrumentation

### DISCUSSION THREAD

**Discussion by amannn** [vercel/next.js#69294] [Ideas]
### Goals

Allow to register client-only code in `instrumentation.ts`.

### Background

[`instrumentation.ts`](https://nextjs.org/docs/app/building-your-application/optimizing/instrumentation) is currently available for code that executes in Node.js and Edge. However, as far as I know, there is currently no equivalent that is called for the client side.

Error-tracking libraries like Sentry currently enable client-side instrumentation, by [wrapping each client-side entry point](https://github.com/getsentry/sentry-javascript/blob/master/packages/nextjs/src/config/loaders/wrappingLoader.ts) and injecting a custom script into the build.

While this works, I'd argue it's somewhat brittle and also requires quite some custom logic to work. Instead, I'd find it very helpful if Next.js would expose a built-in way that covers this.

My particular use case is adapting the [error handling capabilities of `next-intl`](https://next-intl-docs.vercel.app/docs/usage/configuration#error-handling) to register a global error handler that is automatically called in all environments if registered in `instrumentation.ts`.

/cc @lforst in case you're interested in this too

Related: https://github.com/vercel/next.js/issues/57563

### Proposal

```ts
export async function register() {
  // ✨ New
  if (process.env.NEXT_RUNTIME === 'client') {
    await import('./instrumentation-client')
  }

  if (process.env.NEXT_RUNTIME === 'nodejs') {
    await import('./instrumentation-node')
  }
 
  if (process.env.NEXT_RUNTIME === 'edge') {
    await import('./instrumentation-edge')
  }
}
```

**Comment by lforst** [comment]
Random additional input here: It would be crucial to have something similar to the `onRequestError` hook for client-side Pages Router errors. Currently, Next.js simply swallows all the rendering errors. With the Sentry SDK, we apply build-time instrumentation to wrap all Pages in the pages router during the build to report errors that happen during the render cycle. This is quite fragile and also not doable with Turbopack.

**Comment by mderazon** [comment]
Also interested in sending otel traces from client side. Would be nice in Next.js did that as adding tracing to the frontend code is currently complicated:
- https://opentelemetry.io/blog/2022/frontend-overhaul/#opentelemetry-instrumentation
- https://github.com/open-telemetry/opentelemetry-demo/tree/395f38572d4b1352883712cc080d6894f5f9faf8/src/frontend/utils/telemetry

Having end-to-end traces could nice

**Comment by Netail** [comment]
Another use-case; Initializing Grafana Faro

**Comment by Netail** [comment]
Look what we have here; https://github.com/vercel/next.js/pull/76916 👀

**Comment by shirahamax** [comment]
@gaojude

Hello, thank you for adding the clientInstrumentationHook feature :)

I recently upgraded my product to version 15.3 and noticed that clientInstrumentationHook is still listed under ExperimentalConfig.
Would it be possible to remove this flag from ExperimentalConfig?
https://github.com/vercel/next.js/blob/v15.3.0/packages/next/src/server/config-shared.ts#L651

Also, may I ask another question?
The documentation only mentions in the App Router — is the Pages Router not supported?👀
https://nextjs.org/docs/app/api-reference/file-conventions/instrumentation-client

**Reply by gaojude** [reply]
> @gaojude
> 
> 
> 
> Hello, thank you for adding the clientInstrumentationHook feature :)
> 
> 
> 
> I recently upgraded my product to version 15.3 and noticed that clientInstrumentationHook is still listed under ExperimentalConfig.
> 
> Would it be possible to remove this flag from ExperimentalConfig?
> 
> https://github.com/vercel/next.js/blob/v15.3.0/packages/next/src/server/config-shared.ts#L651
> 
> 
> 
> Also, may I ask another question?
> 
> The documentation only mentions in the App Router — is the Pages Router not supported?👀
> 
> https://nextjs.org/docs/app/api-reference/file-conventions/instrumentation-client

Both supported. We should fix the doc. cc @delbaoliveira

**Reply by shirahamax** [reply]
Thank you for your reply. I understand :)

**Comment by polaris-fullstack** [comment]
Thanks for the detailed write-up — I totally agree that having first-class support for client-side instrumentation in instrumentation.ts would simplify things a lot, especially for cross-environment error tracking setups.

At the moment, like you said, most workarounds rely on bundler-level injection or wrapping entry points with tools like Sentry or LogRocket. While they work, they often lead to fragile setups and extra maintenance overhead.

In my case, I’ve been experimenting with something like this:
`if (typeof window !== "undefined") {
  // Simulate a client-side instrumentation hook
  window.addEventListener("error", (e) => {
    // Forward to custom logger or error tracking service
    console.error("Client-side error captured:", e);
  });
}`
Not perfect — and it doesn't tie into instrumentation.ts, but it at least gets global error hooks on the client.

Ideally, Next.js could expose something like:
`export function registerClientInstrumentation() {
  // Called automatically on the client, just like instrumentation.ts is on the server
}
`
This would help unify error handling logic for libraries like next-intl and give more confidence across environments.

Really interested in what the core team thinks here — seems like a great DX improvement.

**Comment by Adebesin-Cell** [comment]
Hmm… I’m trying to use **PostHog** with Next.js 15’s [instrumentation-client.ts](https://nextjs.org/docs/app/api-reference/file-conventions/instrumentation-client).

```ts
import posthog from "posthog-js";

posthog.init(process.env.NEXT_PUBLIC_POSTHOG_KEY!, {
	api_host: "/ingest",
	ui_host: "https://us.posthog.com",
	capture_exceptions: true,
	debug: process.env.NODE_ENV === "development",
});
```

It works fine locally, both dev and build, but during `turbo build` in GitHub workflow, I’m running into Node-related module errors coming from `posthog-js` (like `node:fs` / `node:child_process`).


Has anyone else hit this or found a clean workaround? @amannn 🙏

**Reply by amannn** [reply]
I haven't gotten around to try `instrumentation-client.ts` yet and don't have any experience with it so far—sorry!

**Reply by Adebesin-Cell** [reply]
No worries, thank you! 🙏
Hoping someone else here might have tried it or found a workaround :)

**Reply by tmoennich-nf** [reply]
@Adebesin-Cell FYI we also had issues with the PostHog node library (within Next v16) and switched by to Webpack to fix the imports like this:

```ts
  webpack: (config: WebpackConfiguration) => {
    if (!config.resolve) {
      config.resolve = {};
    }

    config.resolve.fallback = {
      ...config.resolve.fallback,
      fs: false,
      readline: false,
      // @TODO: Maybe no longer needed?
      'node:fs': false,
    };

    // Remove "node:" prefix from imports
    config.plugins?.push(
      new NormalModuleReplacementPlugin(/^node:/, resource => {
        resource.request = resource.request.replace(/^node:/, '');
      })
    );

    return config;
  },
```

I needed a quickfix and therefore didn't checked yet if something similar is possible with Turbopack.
But maybe that already helps 🙏
